src/tests/profiler/gcheapenumeration/gcheapenumeration.cs

```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Profiler.Tests
+{
+    class GCHeapEnumerationTests
+    {
+        static readonly string ShouldSetMonitorGCEventMaskEnvVar = "Set_Monitor_GC_Event_Mask";
+        static readonly Guid GCHeapEnumerationProfilerGuid = new Guid("8753F0E1-6D6D-4329-B8E1-334918869C15");
+        static List<object> _rootObjects = new List<object>();
+
+        [DllImport("Profiler")]
+        private static extern void EnumerateGCHeapObjects();
+
+        [DllImport("Profiler")]
+        private static extern void SuspendRuntime();
+
+        [DllImport("Profiler")]
+        private static extern void ResumeRuntime();
+
+        [DllImport("Profiler")]
+        private static extern void EnumerateHeapObjectsInBackgroundThread();
+
+        public static int EnumerateGCHeapObjectsSingleThreadNoPriorSuspension()
+        {
+            _rootObjects.Add(new CustomGCHeapObject());
+            EnumerateGCHeapObjects();
+            return 100;
+        }
+
+        public static int EnumerateGCHeapObjectsSingleThreadWithinProfilerRequestedRuntimeSuspension()
+        {
+            _rootObjects.Add(new CustomGCHeapObject());
+            SuspendRuntime();
+            EnumerateGCHeapObjects();
+            ResumeRuntime();
+            return 100;
+        }
+
+        public static int EnumerateGCHeapObjectsInBackgroundThreadWithRuntimeSuspension()
+        {
+            _rootObjects.Add(new CustomGCHeapObject());
+            EnumerateHeapObjectsInBackgroundThread();
+            GC.Collect();
+            return 100;
+        }
+
+        public static int EnumerateGCHeapObjectsMultiThreadWithCompetingRuntimeSuspension()
+        {
+            ManualResetEvent startEvent = new ManualResetEvent(false);
+            Thread gcThread = new Thread(() =>
+            {
+                startEvent.WaitOne();
+                GC.Collect();
+            });
+            gcThread.Name = "GC.Collect";
+            gcThread.Start();
+
+            Thread enumerateThread = new Thread(() =>
+            {
+                startEvent.WaitOne();
+                Thread.Sleep(1000);
```

Comment:
I don't think this is testing what you intend it to. The GC.Collect() keeps managed code suspended which means this sleep isn't waking up until the GC.Collect() is finished. By the time EnumerateGCHeapObjects() runs there is no concurrent suspension happening. 

To test concurrent suspends you will need to coordinate it from native code. For example you might do this:
```
GCHeapEnumerationProfiler::GarbageCollectionStarted()
{
    _gcStartSleeping = TRUE;
    // if this new call to EnumerateGCHeapObjects doesn't wait for the GC then it is going to
    // observe _gcStartSleeping == TRUE
    thread_list.emplace_back(std::thread([instance]()
                {
                    instance->ValidateEnumerateGCHeapObjects(S_OK);
                }));
   std::this_thread::sleep_for(std::chrono::seconds(1));
   _gcStartSleeping = FALSE;
}

static BOOL STDMETHODCALLTYPE heap_walk_fn(ObjectID object, void* callbackState)
{
    CallbackState* pState = (CallbackState*)callbackState;
    if(pState->instance._gcStartSleeping)
    {
        //Error, no callbacks expected while the GC collection was going on
    }
    ...
}

HRESULT GCHeapEnumerationProfiler::Shutdown()
{
    // make sure async operations have finished
    for (auto& t : threadList) {
        t.join();
    }
}
```

Reply:
For clarification, this would be testing invoking `ProfToEEInterfaceImpl::EnumerateGCHeapObjects` during a non-profiler requested runtime suspension? As in we are expecting it to hit the `else if` in 
```
    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
    {
        g_profControlBlock.fProfilerRequestedRuntimeSuspend = TRUE;
        ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
        ownEESuspension = TRUE;
    }
    else if (!g_profControlBlock.fProfilerRequestedRuntimeSuspend)
    {
        return CORPROF_E_SUSPENSION_IN_PROGRESS;
    }
    else
    {
        // The profiler requested a runtime suspension before invoking this API,
        // so the responsibility of resuming the runtime is outside the scope of this API.
        // Given that the profiler owns the suspension, walking the GC Heap is safe.
    }
```
and throw the error `CORPROF_E_SUSPENSION_IN_PROGRESS`

Originally, I was intending the [3rd test](https://gist.github.com/mdh1418/3bb38bbb120ab99e6c3f2225aaca041f)
to be the "background native thread invoking EnumerateGCHeapObjects and rejoining a thread that has an ongoing non-profiler requested runtime suspension" test, and for this (4th test) to be multiple managed threads racing for `ThreadSuspend::SuspendEE`, in case the above if block
```
    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
    {
        g_profControlBlock.fProfilerRequestedRuntimeSuspend = TRUE;
        ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
        ownEESuspension = TRUE;
    }
```
is entered before a different native thread invokes `ThreadSuspend::SuspendEE` (which would have toggled the condition to be false)

Reply:
> As in we are expecting it to hit ProfToEEInterfaceImpl::EnumerateGCHeapObjects's else if in

I proposed some alternate logic for how you would check for suspensions. The test case I suggested in my comment above does the EnumerateGCHeapObjects on a different thread than the one which suspended the runtime so I'd expect it to enumerate successfully after waiting for the GC to complete. It should run through this part of the code I recommended above:

```
else
{
    _ASSERTE(!suspendedByThisThread);
    // Its possible some background threads are suspending and resuming the runtime
    // concurrently. We need to suspend the runtime on this thread to be certain the heap
```

Reply:
Ah I missed the above recommendations for checking thread suspension within `ProfToEEInterfaceImpl::EnumerateGCHeapObjects`, after implementing a test as you suggested in the first comment in this thread, I attached a debugger and the background thread that invokes EnumerateGCHeapObjects does indeed get blocked on `ThreadSuspend::SuspendEE` until GC completes
```
00 0000001b`0657f218 00007ffb`06d640e3     ntdll!ZwWaitForAlertByThreadId+0x14
01 0000001b`0657f220 00007ffb`06d51ce4     ntdll!TpWorkOnBehalfClearTicket+0x2c3
02 0000001b`0657f330 00007ffb`06d51ad2     ntdll!RtlEnterCriticalSection+0x254
03 0000001b`0657f3b0 00007ffa`5b26652e     ntdll!RtlEnterCriticalSection+0x42
04 0000001b`0657f3e0 00007ffa`5b51372b     coreclr!CrstBase::Enter+0x27e [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\crst.cpp @ 325] 
05 0000001b`0657f440 00007ffa`5b8abd8b     coreclr!ThreadStore::Enter+0x1bb [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\threads.cpp @ 5092] 
06 0000001b`0657f560 00007ffa`5b8b18ad     coreclr!ThreadSuspend::LockThreadStore+0x25b [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\threadsuspend.cpp @ 1882] 
07 0000001b`0657f690 00007ffa`5b808f13     coreclr!ThreadSuspend::SuspendEE+0x12d [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\threadsuspend.cpp @ 5559] 
08 0000001b`0657f770 00007ffa`672cd542     coreclr!ProfToEEInterfaceImpl::EnumerateGCHeapObjects+0x473 [C:\Users\mihw\source\repos\mdh1418\runtime\src\coreclr\vm\proftoeeinterfaceimpl.cpp @ 7707] 
09 0000001b`0657f8f0 00007ffa`672cdda8     Profiler!GCHeapEnumerationProfiler::EnumerateGCHeapObjects+0xa2 [C:\Users\mihw\source\repos\mdh1418\runtime\src\tests\profiler\native\gcheapenumerationprofiler\gcheapenumerationprofiler.cpp @ 186] 
0a 0000001b`0657f960 00007ffa`672cd335     Profiler!GCHeapEnumerationProfiler::ValidateEnumerateGCHeapObjects+0x18 [C:\Users\mihw\source\repos\mdh1418\runtime\src\tests\profiler\native\gcheapenumerationprofiler\gcheapenumerationprofiler.cpp @ 190] 
0b 0000001b`0657f9a0 00007ffa`672ccab4     Profiler!<lambda_e2419102f21665b3332b03f113586c25>::operator()+0x25 [C:\Users\mihw\source\repos\mdh1418\runtime\src\tests\profiler\native\gcheapenumerationprofiler\gcheapenumerationprofiler.cpp @ 74] 
0c 0000001b`0657f9d0 00007ffa`672cbe19     Profiler!std::invoke<<lambda_e2419102f21665b3332b03f113586c25> >+0x14 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\type_traits @ 1705] 
0d 0000001b`0657fa00 00007ffa`673b63a0     Profiler!std::thread::_Invoke<std::tuple<<lambda_e2419102f21665b3332b03f113586c25> >,0>+0x59 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\thread @ 61] 
0e 0000001b`0657fa60 00007ffb`055f257d     Profiler!thread_start<unsigned int (__cdecl*)(void *),1>+0xb0 [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 97] 
0f 0000001b`0657fac0 00007ffb`06d8af28     KERNEL32!BaseThreadInitThunk+0x1d
10 0000001b`0657faf0 00000000`00000000     ntdll!RtlUserThreadStart+0x28
```

Reply:
Modified the test case to simulate scenario 1, where the thread suspending runtime through GC spins off another native thread to invoke EnumerateGCHeapObjects.

Will add the test case to simulate scenario 2 if I can find the test attribute that bypasses contract violation assert and hits the expected errors.